app/models/transformation_mapping.rb

```diff
@@ -0,0 +1,64 @@
+class TransformationMapping < ApplicationRecord
+  has_many :transformation_mapping_items, :dependent => :destroy
+
+  validates :name, :presence => true, :uniqueness => true
+
+  virtual_column :mappings, :type => :string
+
+  # forward compatible: next release the table will be polymorphic
+  virtual_column :type,     :type => :string
+
+  def self.inheritance_column
+    "none"
+  end
+
+  def type
+    "MigrationInfrastructureMapping"
+  end
+
+  def type=(type)
+  end
+
+  # Assumption: source and destination types are the same. Works for infrastructure mapping
+  # group mapping items first by destination_type, then by destination_id
+  # simplify records with IDs.
+  # Sample result (Hash with key being the underscore case of the model)
+  #
+  # ems_cluster:
+  #   - destination: "2":
+  #     sources: ["100", "200"]
+  #   - destination: "3"
+  #     sources: ["301", "302"]
+  # network:
+  #   - destination: "2"
+  #     sources: ["202", "204"]
+  # storage:
+  #   - destination: "4"
+  #     sources: ["33", "32"]
+  def mappings
+    dst_type_groups = transformation_mapping_items.group_by(&:destination_type)
+    dst_type_groups.each_with_object({}) do |(dst_type, items), results|
+      dst_groups = items.group_by { |item| item.destination_id.to_s }
+      results[dst_type.underscore] = dst_groups.each_with_object({}) do |(dst_id, item_array), dst_id_groups|
+        dst_id_groups["destination"] = dst_id
+        dst_id_groups["sources"] = item_array.collect { |item| item.source_id.to_s }
+      end
+    end
+  end
+
+  def mappings=(options)
+    options.each do |dst_type, item_map_array|
+      item_map_array.each do |item_map|
+        destination_id = item_map['destination']
+        item_map['sources'].each do |source_id|
+          transformation_mapping_items << TransformationMappingItem.new(
+            :source_id        => source_id,
+            :source_type      => dst_type.camelize,
```

Comment:
Instead of camelizing, I think that the API should likely deconstruct the type from the api collection to the actual `source_type`. I suggest this because not all of the source types will align perfectly. I am not sure all of the types a transformation item will allow, but just as an example, `providers` aligns to `ExtManagementSystem`.  I think it would be best for the API to transform `providers` from the use to `ExtManagementSystem`. Thoughts?

Reply:
I am perfectly fine with what you proposed.